File: src/controllers/roomController.js
const db = require('../db');

exports.getAllRooms = async (req, res) => {
  try {
    const result = await db.query("SELECT * FROM rooms ORDER BY id ASC");
    const rooms = result.rows.map(r => ({ ...r, amenities: JSON.parse(r.amenities) }));
    res.json(rooms);
  } catch(err) {
    res.status(500).json({ error: err.message });
  }
};

exports.getAvailableRooms = async (req, res) => {
  try {
    const { date, time, duration } = req.query;
    
    const result = await db.query("SELECT * FROM rooms ORDER BY id ASC");
    const allRooms = result.rows.map(r => ({ ...r, amenities: JSON.parse(r.amenities) }));

    let filterDate = date;
    let filterStart = null;
    let filterEnd = null;

    if (date && time && duration) {
      filterStart = parseInt(time.split(':')[0]) * 60 + parseInt(time.split(':')[1]);
      filterEnd = filterStart + parseInt(duration) * 60;
    } else {
      const now = new Date();
      filterDate = filterDate || now.toISOString().split('T')[0];
      const hours = now.getHours();
      const minutes = now.getMinutes();
      filterStart = hours * 60 + minutes;
      filterEnd = filterStart + 60; 
    }

    const bookingsRes = await db.query("SELECT room_id, start_time, duration FROM bookings WHERE date = $1", [filterDate]);
    const dayBookings = bookingsRes.rows;
    
    const unavailableRoomIds = new Set();
    
    dayBookings.forEach(b => {
      if(!b.start_time || !b.duration) return;
      const bStart = parseInt(b.start_time.split(':')[0]) * 60 + parseInt(b.start_time.split(':')[1]);
      const bEnd = bStart + parseInt(b.duration) * 60;
      
      if (filterStart < bEnd && filterEnd > bStart) {
        unavailableRoomIds.add(b.room_id);
      }
    });

    const availableRooms = allRooms.filter(r => !unavailableRoomIds.has(r.id));
    res.json({ available_rooms_count: availableRooms.length, available_rooms: availableRooms });
  } catch(err) {
    res.status(500).json({ error: err.message });
  }
};

File: controllers/room_controller.py
import json
from datetime import datetime
from flask import request, jsonify
from db import get_connection

def get_all_rooms():
    """GET /rooms - Returns all rooms from the database."""
    conn = get_connection()
    cur = conn.cursor()
    try:
        cur.execute("SELECT * FROM rooms ORDER BY id ASC")
        columns = [desc[0] for desc in cur.description]
        rooms = [dict(zip(columns, row)) for row in cur.fetchall()]
        for room in rooms:
            room["amenities"] = json.loads(room["amenities"])
        return jsonify(rooms)
    except Exception as e:
        return jsonify({"error": str(e)}), 500
    finally:
        cur.close()
        conn.close()

def get_available_rooms():
    """GET /rooms/available - Returns rooms not booked in given window."""
    conn = get_connection()
    cur = conn.cursor()
    try:
        cur.execute("SELECT * FROM rooms ORDER BY id ASC")
        columns = [desc[0] for desc in cur.description]
        all_rooms = [dict(zip(columns, row)) for row in cur.fetchall()]
        for room in all_rooms:
            room["amenities"] = json.loads(room["amenities"])

        date, time_str, duration, loc_param = request.args.get("date"), request.args.get("time"), request.args.get("duration"), request.args.get("location")
        if date and time_str and duration:
            h, m = map(int, time_str.split(":"))
            filter_start, filter_end = h * 60 + m, (h * 60 + m) + int(duration) * 60
            filter_date = date
        else:
            now = datetime.now()
            filter_date = date or now.strftime("%Y-%m-%d")
            filter_start = now.hour * 60 + now.minute
            filter_end = filter_start + 60 

        cur.execute("SELECT room_id, start_time, duration FROM bookings WHERE date = %s", (filter_date,))
        unavailable_ids = {
            b[0] for b in cur.fetchall() if b[1] and b[2] and 
            filter_start < (int(b[1].split(":")[0]) * 60 + int(b[1].split(":")[1]) + int(b[2]) * 60) and 
            filter_end > (int(b[1].split(":")[0]) * 60 + int(b[1].split(":")[1]))
        }

        available_rooms = [r for r in all_rooms if r["id"] not in unavailable_ids]
        if loc_param:
            available_rooms = [r for r in available_rooms if r["location"].lower() == loc_param.lower()]

        return jsonify({"available_rooms_count": len(available_rooms), "available_rooms": available_rooms})
    except Exception as e:
        return jsonify({"error": str(e)}), 500
    finally:
        cur.close()
        conn.close()
